reference/src/glossary.md

```diff
@@ -56,6 +55,35 @@ somewhat differently from this definition. However, that's considered a low
 level detail of a particular Rust implementation. When programming Rust, the
 Abstract Rust Machine is intended to operate according to the definition here.
 
+#### (Pointer) Provenance
+
+The *provenance* of a pointer is used to distinguish pointers that point to the same memory address (i.e., pointers that, when cast to `usize`, will compare equal).
+Provenance is extra state that only exists in the Rust Abstract Machine; it is needed to specify program behavior but not present any more when the program runs on real hardware.
+
+For example, we have to distinguish pointers to the same location if they originated from different allocations.
+Cross-allocation pointer arithmetic [does not lead to usable pointers](https://doc.rust-lang.org/std/primitive.pointer.html#method.wrapping_offset), so the Rust Abstract Machine *somehow* has to remember the original allocation to which a pointer pointed.
+It uses provenance to achieve this:
+
+```rust
+// Let's assume the two allocations here have base addresses 0x100 and 0x200.
+// We write pointer provenance as `@N` where `N` is some kind of ID uniquely
+// identifying the allocation.
+let raw1 = Box::into_raw(Box::new(13u8));
+let raw2 = Box::into_raw(Box::new(42u8));
+let raw2_wrong = raw1.wrapping_add(raw2.wrapping_sub(raw1 as usize) as usize);
+// These pointers now have the following values:
+// raw1 points to address 0x100 and has provenance @1.
+// raw2 points to address 0x200 and has provenance @2.
+// raw2_wrong points to address 0x200 and has provenance @1.
+// In other words, raw2 and raw2_wrong have same *address*...
+assert_eq!(raw2 as usize, raw2_wrong as usize);
+// ...but it would be UB to use raw2_wrong, as it was obtained by
```

Comment:
Can one offset back raw2_wrong to raw1 and use raw2_wrong to access the memory of the allocation 1 ?

Reply:
If that the case, what is not allowed is to use a pointer to one allocation to access the memory of another allocation.

Reply:
Yes. You can use `wrapping_offset` to move the pointer back to address `0x100`, and it still has provenance `@1`, which matches the allocation that actually sits at address `0x100`, so the pointer is good.

The issue with `raw2_wrong` is that the allocation at address `0x200` has ID `@2`, but the pointer has provenance `@1`. This mismatch is used, by the Abstract Machine, to declare dereferencing `raw2_wrong` as UB.

Reply:
> If that the case, what is not allowed is to use a pointer to one allocation to access the memory of another allocation

Nit: a pointer *originating* from one allocation. "Pointer to an allocation" is a fuzzy term.

> cross-allocation arithmetic wouldn't be UB per se.

Ah, I should clarify. The arithmetic is not UB but the pointer isnt usable for other allocations.

Reply:
Exactly, I would just say that since the pointer has provenance `@1`, it cannot be used to access any memory that does not correspond to the `@1` allocation. So dereferencing it to access the memory of the allocation `@2` is undefined.

Reply:
Like this? I changed the text.

Reply:
> ...but it would be UB to use raw2_wrong, as it has the wrong *provenance*:

I would at "...but it would be UB to use raw2_wrong to access @2, as ...", since using it in other ways is ok.

Reply:
I made it "UB to dereference raw2_wrong".

Reply:
Sounds good, thanks!